Topics we will look here

sentence tokenization

word tokenization

lemmetization

TF-IDF



In [9]:
import nltk
nltk.download('punkt')

data ="India is the land of people of diverse religions, traditions and castes. People of India represent the essence of our country. The diverse colours of India are its cultures, traditions and religions."

sentences = nltk.sent_tokenize(data)

print(sentences)

['India is the land of people of diverse religions, traditions and castes.', 'People of India represent the essence of our country.', 'The diverse colours of India are its cultures, traditions and religions.']


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Supratik\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [11]:
for i in range(len(sentences)):
    print(nltk.word_tokenize(sentences[i]))


['India', 'is', 'the', 'land', 'of', 'people', 'of', 'diverse', 'religions', ',', 'traditions', 'and', 'castes', '.']
['People', 'of', 'India', 'represent', 'the', 'essence', 'of', 'our', 'country', '.']
['The', 'diverse', 'colours', 'of', 'India', 'are', 'its', 'cultures', ',', 'traditions', 'and', 'religions', '.']


In [12]:
nltk.word_tokenize(data)

['India',
 'is',
 'the',
 'land',
 'of',
 'people',
 'of',
 'diverse',
 'religions',
 ',',
 'traditions',
 'and',
 'castes',
 '.',
 'People',
 'of',
 'India',
 'represent',
 'the',
 'essence',
 'of',
 'our',
 'country',
 '.',
 'The',
 'diverse',
 'colours',
 'of',
 'India',
 'are',
 'its',
 'cultures',
 ',',
 'traditions',
 'and',
 'religions',
 '.']

In [13]:
from nltk.corpus import stopwords

stop_words = stopwords.words('english')
stop_words

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [14]:
import string

punct = string.punctuation
print(punct)

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


In [15]:
# cleaning data

clean_data = []


for word in nltk.word_tokenize(data):
    if word not in punct:
        if word not in stop_words:
            clean_data.append(word)

clean_data

['India',
 'land',
 'people',
 'diverse',
 'religions',
 'traditions',
 'castes',
 'People',
 'India',
 'represent',
 'essence',
 'country',
 'The',
 'diverse',
 'colours',
 'India',
 'cultures',
 'traditions',
 'religions']

In [17]:
from nltk.stem import WordNetLemmatizer

lemma = WordNetLemmatizer()

for word in clean_data:
    print(lemma.lemmatize(word))

India
land
people
diverse
religion
tradition
caste
People
India
represent
essence
country
The
diverse
colour
India
culture
tradition
religion


In [18]:
from nltk.stem import PorterStemmer

ps = PorterStemmer()

for word in clean_data:
    print(ps.stem(word))

india
land
peopl
divers
religion
tradit
cast
peopl
india
repres
essenc
countri
the
divers
colour
india
cultur
tradit
religion


In [20]:
from nltk.stem import SnowballStemmer #known as porter2, recently popular

porter = SnowballStemmer('english')

for word in clean_data:
    print(porter.stem(word))

india
land
peopl
divers
religion
tradit
cast
peopl
india
repres
essenc
countri
the
divers
colour
india
cultur
tradit
religion


**Term Frequency :**
It is simply the frequency in which a word appears in a document in comparison to the total number words in the document. Mathematically given as:

Term frequency = (Number of times a word appears in the document) / (Total number of words in the document)

**Inverse Document Frequency:**
Term frequency has a disadvantage that it tends to give higher weights to words with higher frequency. In such cases words like ‘a’, ‘the’, ‘in’, ’of’ etc. appears more in the documents than other regular words. Thus, more important words are wrongly given lower weights as their frequency is less. To tackle this problem IDF was introduced. IDF decreases the weights of such high frequency terms and increases the weight of terms with rare occurrence. Mathematically it is given as:

Inverse Document Frequency = log [(Number of documents)/(Number of documents the word appears in)]

note: [log has base 2]

Let us take the below example

In [21]:
doc = ["I am a boy.","We are playing","I saw a girl"]
print(doc)

['I am a boy.', 'We are playing', 'I saw a girl']


<img src = "tfidf1.jpg">

In the above table, we have calculated the term frequency as well as inverse document frequency of each of the words present in the 3 documents given. 

Now, let's calculate the tf-idf score for each term. Since, words of one document is not present in another document, we will have tf-idf value 0 for them e.g. words of doc1 will have 0 tf-idf for doc2 and doc3.

<img src = "tfidf2.jpg">

Now we will see the same example using python.

In [23]:
import pandas as pd

In [25]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()

doc_vector_tfidf = tfidf.fit_transform(doc)
#print(tfidf.get_feature_names())
#print(doc_vector_tfidf.todense())
df = pd.DataFrame(doc_vector_tfidf.todense(),columns=tfidf.get_feature_names())
df

am      are       boy      girl  playing       saw       we
0  0.707107  0.00000  0.707107  0.000000  0.00000  0.000000  0.00000
1  0.000000  0.57735  0.000000  0.000000  0.57735  0.000000  0.57735
2  0.000000  0.00000  0.000000  0.707107  0.00000  0.707107  0.00000

Just to note, in this above table you won't find words like 'I' , 'a', because these are stop words and in the above document their importance is very very less. Remember one fact, **Greater the TF-IDF score, the word has more importance in the paragraph or in that series of document.**

Now we will do the spam ham classification using TFIDF 

In [27]:
messages = pd.read_table('smsspamcollection/SMSSpamCollection', names = ['label','message'])
messages.head()

label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [28]:
import re

In [58]:
messages.shape

(5572, 2)

In [59]:
len(messages)

5572

In [65]:
corpus = []

for i in range(0,len(messages)):
    review = re.sub('[^a-zA-Z]',' ',messages['message'][i])
    review = review.lower()
    review = review.split()
    
    review = [ps.stem(word) for word in review if word not in stop_words]
    review = ' '.join(review)
    corpus.append(review)
    

In [66]:
corpus[1]

'ok lar joke wif u oni'

In [67]:
len(corpus)

5572

In [68]:
y = messages['label']
y.head()

0     ham
1     ham
2    spam
3     ham
4     ham
Name: label, dtype: object

In [69]:
y = pd.get_dummies(y)
y.head()

ham  spam
0    1     0
1    1     0
2    0     1
3    1     0
4    1     0

In [70]:
type(y)

pandas.core.frame.DataFrame

In [71]:
y = y.drop('ham',axis = 1)

In [72]:
y.head()

spam
0     0
1     0
2     1
3     0
4     0

In [86]:
# implementing TFIDF on the given data

from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(max_features=5000) # used max_features to increase the performance

X = tfidf.fit_transform(corpus)



In [87]:
type(X)

scipy.sparse.csr.csr_matrix

In [88]:
reformed_data = pd.DataFrame(X.todense(),columns=tfidf.get_feature_names())

reformed_data.head()

aa  aah  aaniy  aaooooright  aathi   ab  abbey  abdomen  abeg  abel  ...  \
0  0.0  0.0    0.0          0.0    0.0  0.0    0.0      0.0   0.0   0.0  ...   
1  0.0  0.0    0.0          0.0    0.0  0.0    0.0      0.0   0.0   0.0  ...   
2  0.0  0.0    0.0          0.0    0.0  0.0    0.0      0.0   0.0   0.0  ...   
3  0.0  0.0    0.0          0.0    0.0  0.0    0.0      0.0   0.0   0.0  ...   
4  0.0  0.0    0.0          0.0    0.0  0.0    0.0      0.0   0.0   0.0  ...   

   zebra  zed  zero   zf  zhong  zindgi  zoe  zogtoriu  zoom  zouk  
0    0.0  0.0   0.0  0.0    0.0     0.0  0.0       0.0   0.0   0.0  
1    0.0  0.0   0.0  0.0    0.0     0.0  0.0       0.0   0.0   0.0  
2    0.0  0.0   0.0  0.0    0.0     0.0  0.0       0.0   0.0   0.0  
3    0.0  0.0   0.0  0.0    0.0     0.0  0.0       0.0   0.0   0.0  
4    0.0  0.0   0.0  0.0    0.0     0.0  0.0       0.0   0.0   0.0  

[5 rows x 5000 columns]

In [89]:
reformed_data.shape

(5572, 5000)

In [90]:
X = reformed_data

In [91]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2, random_state = 33)


In [92]:
X_train.shape

(4457, 5000)

In [93]:
X_test.shape

(1115, 5000)

In [94]:
from sklearn.naive_bayes import MultinomialNB

spam_detection_model = MultinomialNB()

spam_detection_model.fit(X_train,y_train)



C:\Users\Supratik\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [95]:
y_pred = spam_detection_model.predict(X_test)

In [96]:
from sklearn.metrics import accuracy_score,confusion_matrix

model_accuracy_score = accuracy_score(y_test,y_pred)


In [97]:
model_accuracy_score

0.9775784753363229

In [98]:
cm = confusion_matrix(y_test,y_pred)
cm

array([[962,   0],
       [ 25, 128]], dtype=int64)

Hyperparameter tuning for MultinomialNB

In [99]:
alpha = 0.2

import numpy as np

In [100]:
for alpha in np.arange(0.2,1,0.1):
    new_model = MultinomialNB(alpha=alpha)
    new_model.fit(X_train,y_train)
    y_pred_tuned = new_model.predict(X_test)
    score = accuracy_score(y_test,y_pred_tuned)
    print("Alpha : {}, Score : {}" .format(alpha,score))

C:\Users\Supratik\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Supratik\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Alpha : 0.2, Score : 0.9865470852017937
Alpha : 0.30000000000000004, Score : 0.9847533632286996


C:\Users\Supratik\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Supratik\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Alpha : 0.4000000000000001, Score : 0.9838565022421525
Alpha : 0.5000000000000001, Score : 0.9820627802690582


C:\Users\Supratik\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Supratik\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Alpha : 0.6000000000000001, Score : 0.979372197309417
Alpha : 0.7000000000000002, Score : 0.97847533632287


C:\Users\Supratik\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Supratik\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Alpha : 0.8000000000000003, Score : 0.97847533632287
Alpha : 0.9000000000000001, Score : 0.9766816143497757


Here, from above we can conclude that the best alpha value for this model is 0.2, with that alpha value the model score is  98.65%. So, while creating our model we can use **spam_detection_model = MultinomialNB(alpha = 0.2)**